### PR TITLE
diagnostics: optimize tracing

### PR DESCRIPTION
**Pull request by daixiang0**
Reduce sampler assignment times and supprot all service name of Invoke.

# Description

Reduce sampler assignment times and supprot all service name of Invoke.

## Issue reference

Please reference the issue this PR will close: #_[issue number]_

## Checklist

Please make sure you've  completed the relevant tasks for this PR, out of the following list:

* [x] Code compiles correctly
* [ ] Created/updated tests
* [X] Unit tests passing
* [x] End-to-end tests passing
* [ ] Extended the documentation / Created issue in the https://github.com/dapr/docs/ repo: dapr/docs#_[issue number]_
* [ ] Specification has been updated / Created issue in the https://github.com/dapr/docs/ repo: dapr/docs#_[issue number]_
* [ ] Provided sample for the feature / Created issue in the https://github.com/dapr/docs/ repo: dapr/docs#_[issue number]_

RELEASE NOTE: **ENABLED** reduced sampler assignment times and support all service name of Invoke.

### CODE REVIEW

**Comment by daixiang0** [comment]
ping @artursouza

**Comment by yaron2** [comment]
@artursouza can you review this?

File: pkg/diagnostics/grpc_tracing.go [hunk 1]
```diff
@@ -51,13 +51,31 @@ func GRPCTraceUnaryServerInterceptor(appID string, spec config.TracingSpec) grpc
 			spanKind = trace.WithSpanKind(trace.SpanKindClient)
 		}
 
-		ctx, span = trace.StartSpanWithRemoteParent(ctx, spanName, sc, sampler, spanKind)
+		ctx, span = trace.StartSpanWithRemoteParent(ctx, info.FullMethod, sc, sampler, spanKind)
+		if span.SpanContext().IsSampled() {
```
Comment (artursouza) [inline pkg/diagnostics/grpc_tracing.go:55]: Please, use a variable for this and reuse the same variable in the following `if`, this way both `ifs` are either on or off and not mixed.

File: pkg/runtime/runtime.go [hunk 2]
```diff
@@ -710,7 +710,7 @@ func (a *DaprRuntime) sendBindingEventToApp(bindingName string, data []byte, met
 		if span != nil {
 			m := diag.ConstructInputBindingSpanAttributes(
 				bindingName,
-				"/dapr.proto.runtime.v1.AppCallback/OnBindingEvent")
+				fmt.Sprintf("/%s/OnBindingEvent", runtimev1pb.AppCallback_ServiceDesc.ServiceName))
```
Comment (artursouza) [inline pkg/runtime/runtime.go:713]: I would keep the constant in favor of avoiding string allocation.

File: pkg/diagnostics/grpc_tracing.go [hunk 3]
```diff
@@ -51,13 +51,31 @@ func GRPCTraceUnaryServerInterceptor(appID string, spec config.TracingSpec) grpc
...
+
+			// Correct the span name based on API.
+			if sname, ok := reqSpanAttr[daprAPISpanNameInternal]; ok {
+				span.SetName(sname)
+			}
+		}
 
 		// Add grpc-trace-bin header for all non-invocation api's
-		if info.FullMethod != "/dapr.proto.runtime.v1.Dapr/InvokeService" {
+		target := fmt.Sprintf("/%s/InvokeService", runtimev1pb.Dapr_ServiceDesc.ServiceName)
```
Comment (artursouza) [inline pkg/diagnostics/grpc_tracing.go:77]: I would keep the constant in favor of avoiding string allocation.

**Review by artursouza — CHANGES_REQUESTED** [review]
LGTM, just some minor changes.

File: pkg/diagnostics/grpc_tracing.go (hunk 3 above)
Comment (daixiang0) [reply pkg/diagnostics/grpc_tracing.go]: Sorry for force push :(

**Comment by daixiang0** [comment]
ping @artursouza

**Review by artursouza — DISMISSED** [review]

**Comment by daixiang0** [comment]
@artursouza  does perf test hung?

**Review by artursouza — APPROVED** [review]

**Pull request merged**